src/cc/api/BPF.cc

```diff
@@ -559,7 +559,7 @@ std::string BPF::get_uprobe_event(const std::string& binary_path,
 StatusTuple BPF::detach_kprobe_event(const std::string& event,
                                      open_probe_t& attr) {
   if (attr.reader_ptr) {
-    perf_reader_free(attr.reader_ptr);
+    perf_reader_free(attr.reader_ptr, false);
```

Comment:
For C++, maybe we have these individual detach helpers to `perf_reader_free(..., false)` but the one called in destructor to do `perf_reader_free(..., true)`?

Reply:
Right. We do not need to do munmap if we know for sure it is exiting!

Originally, I thought about whether we could return from callback function with indication it wants to exit. There are the following concern:
   . user program may catch Ctrl^C. If this is caught in the middle of perf_read, and the python keyboard exception call exit(), then the hang is not resolved. Is this possible? Should we take care of this?

For the question to call perf_reader_free(..., true) in destructor, we still need to call perf_reader_free(..., false) since the program is not exiting. This still needed for proper
sync between reader and free in normal cases.

Reply:
For
> Actually, in that case, we don't even need to do the munmap since the Process is going away, right?

Just checked. We can skip munmap, but perf_reader_free have to be called to close reader_fd. Otherwise, we are not able to clean the entry in `/sys/kernel/debug/tracing/kprobe_events`.

With debugfs still in use for k/uprobes, we may still need to preserve the current execution path, i.e., during atexit, perf_reader_free needs to be properly called.